### DISCUSSION TITLE
Understanding your GHES graphs part 1 - System Health, Processes, and Authentication

### DISCUSSION THREAD

**Discussion by djdefi** [community/community#42146] [Enterprise]
GitHub Enterprise Servers collect metrics from its services and the base Linux operating system. This data is useful for troubleshooting performance issues and for understanding how your system is being used. The metrics are collected and stored in a time series database, and are displayed in the [GitHub Enterprise Management Console](https://docs.github.com/enterprise-server/admin/configuration/configuring-your-enterprise/accessing-the-management-console). The Management Console includes a [Monitor dashboard](https://docs.github.com/enterprise-server/admin/enterprise-management/monitoring-your-appliance/accessing-the-monitor-dashboard) located at `http(s)://[hostname]/setup/monitor`. This dashboard displays graphs created with data gathered by the built in `collectd` service. Data used in the graphs is sampled every 10 seconds.

Each graph has an informational tooltip describing the graph, which is accessible by hovering over or clicking on the  **_i_**  in the upper left corner of each graph.

Graph data can also be forwarded to an external receiver, by enabling [Collectd forwarding](https://docs.github.com/enterprise-server/admin/enterprise-management/monitoring-your-appliance/configuring-collectd) within the GitHub Enterprise Management Console. This allows building customized dashboards and alerts for your GitHub Enterprise graph data.

This article series will explore what each of the dashboard sections cover and what specific graph trends to watch out for. As each GitHub Enterprise system is unique in user patterns and integrations, we encourage administrators to reach out to the [GitHub support team](https://docs.github.com/enterprise-server/support/contacting-github-support) to assist with interpreting your specific instance's monitor graphs if questions arise. The graph data is included within appliance [Support Bundles](https://docs.github.com/enterprise-server/support/contacting-github-support/providing-data-to-github-support#creating-and-sharing-support-bundles) which can be shared with our support team.

### System Health

The system health graphs provide a general overview of services and system resource utilization. [CPU, Memory, and Load Average](https://docs.github.com/enterprise-server/admin/enterprise-management/monitoring-your-appliance/recommended-alert-thresholds) graphs are useful for identifying trends or times where provisioned resource saturation has occurred.

#### CPU

[![graph-cpu](https://user-images.githubusercontent.com/3662109/44178016-4572ce80-a0a5-11e8-8d58-13050b04d5fb.png)](https://user-images.githubusercontent.com/3662109/44178016-4572ce80-a0a5-11e8-8d58-13050b04d5fb.png)

- [Abnormally high CPU utilization, or prolonged spikes can mean your instance is under-provisioned](https://docs.github.com/enterprise-server/admin/enterprise-management/monitoring-your-appliance/recommended-alert-thresholds#monitoring-cpu-and-load-average-usage).
- In the above example, the CPU was nearly 100% consumed by `user` for a period of time.
- Presence of CPU "steal" time on the CPU graph can be an indication that other virtual machines running on the same host system are saturating the underlying resources, causing the GitHub Enterprise system to wait for CPU cycles.
- User and System are generally the largest consumers of CPU time.

#### Memory

[![graph-memory](https://user-images.githubusercontent.com/3662109/44178015-44da3800-a0a5-11e8-8958-055fcade61ae.png)](https://user-images.githubusercontent.com/3662109/44178015-44da3800-a0a5-11e8-8958-055fcade61ae.png)

- The Linux Kernel provides a layer of in memory disk caching, which is represented by the "cache" on the graph. It is perfectly normal, and recommended to have at least a few GB of cache overhead. The system will attempt to cache as much as possible, but applications can take this memory on demand. Because of this, we consider the total amount of available memory to be the sum of "cached" and "free" values.
- [Running out of available free + cached memory can lead to out of memory (OOM) events](https://docs.github.com/enterprise-server/admin/enterprise-management/monitoring-your-appliance/recommended-alert-thresholds#monitoring-memory-usage), causing services to terminate and unexpected application behavior.

#### Load

[![graph-load](https://user-images.githubusercontent.com/3662109/44178014-44da3800-a0a5-11e8-81be-4f6b9deb7e3c.png)](https://user-images.githubusercontent.com/3662109/44178014-44da3800-a0a5-11e8-81be-4f6b9deb7e3c.png)

- System Load Average is a measurement showing the running task demand on the system.
- [We recommend monitoring the fifteen minute `longterm` system load average](https://docs.github.com/enterprise-server/admin/enterprise-management/monitoring-your-appliance/recommended-alert-thresholds#monitoring-cpu-and-load-average-usage) for values nearing or exceeding the number of CPU cores allocated to the virtual machine.
- When the load average rises above the number of CPU cores, it generally means that tasks are needing to wait for resources before they can run.
- Assuming the above example graph is a GitHub Enterprise system with 2 CPU cores, we can determine that processes are often waiting for resources.

#### Processes

[![graph-proc-run](https://user-images.githubusercontent.com/3662109/44178012-44da3800-a0a5-11e8-898c-abe64c635449.png)](https://user-images.githubusercontent.com/3662109/44178012-44da3800-a0a5-11e8-898c-abe64c635449.png)

- By clicking on `running` in the legend at the bottom of the graph, we can isolate different process states. In the above example we have selected `running` processes.
- The running process count will fluctuate with system activity. Sharp changes or drops could be expected depending on usage trends.
- Large or consistent numbers of `blocked` or `zombie` processes may indicate a service problem.
- It is expected to have processes in the `sleeping` state during normal operation.

#### Files

[![graph-files](https://user-images.githubusercontent.com/3662109/44178011-44da3800-a0a5-11e8-8a51-afd426f42c3b.png)](https://user-images.githubusercontent.com/3662109/44178011-44da3800-a0a5-11e8-8a51-afd426f42c3b.png)

- This graph represents the `max` number of open files, as well as the current number of `used` open files.
- On a healthy system, the number of `used` files should never reach the `max` value. Reaching the `max` can indicate problems with a GitHub Enterprise service.
- Limiting maximum open files is a protection built into Linux to prevent runaway processes from impacting other services on the system.

#### Forks

[![graph-forks](https://user-images.githubusercontent.com/3662109/44178010-4441a180-a0a5-11e8-880d-4dd4497280a4.png)](https://user-images.githubusercontent.com/3662109/44178010-4441a180-a0a5-11e8-880d-4dd4497280a4.png)

- The `fork_rate` trend greatly depends on system activity, and will reach values upwards of 1000-2000 on busy systems.
- Large spikes beyond the observed averages should be investigated.
- This metrics is related to Linux process forking, and is not related to Forking a repository in the GitHub application.

## Processes

The processes graph section looks deeper into the major individual services which make up the GitHub Enterprise appliance. Looking at these services individually can show how usage trends impact system resources over time.

#### Processes

[![Processes graph](https://user-images.githubusercontent.com/3662109/208490739-c2d60969-ce34-4859-835b-1e88b8f554e9.png)](https://user-images.githubusercontent.com/3662109/208490739-c2d60969-ce34-4859-835b-1e88b8f554e9.png)

- Process counts will fluctuate with usage trends.

#### Memory

[![Memory graph](https://user-images.githubusercontent.com/3662109/208491141-bd5d5060-d34c-4f22-9d15-b5b089a9b52a.png)](https://user-images.githubusercontent.com/3662109/208491141-bd5d5060-d34c-4f22-9d15-b5b089a9b52a.png)

- The `unicorn` and `aqueduct` process groups normally consume the highest amount of memory, followed by `memcached`, `mysql`, and `elasticsearch`.
- `unicorn`, `babeld`, and `git-daemon` processes are most influenced by user activity.
- Elements such the size of repositories interacted with and frequency of requests. Because of this, these process graphs can have peaks and valleys.

#### CPU (Kernel)

[![CPU (Kernel) graph](https://user-images.githubusercontent.com/3662109/208491220-05c741cc-0a4b-4443-bd39-f2bbc579d86d.png)](https://user-images.githubusercontent.com/3662109/208491220-05c741cc-0a4b-4443-bd39-f2bbc579d86d.png)

- CPU time consumed by processes accessing hardware directly, via trusted lower level operating system functions.
- The `unicorn` process normally consumes the most CPU time.
- Often has lower values than the following CPU (Application) graph.

#### CPU (Application)

[![CPU (Application) graph](https://user-images.githubusercontent.com/3662109/208491336-54952b08-237e-4ef4-ad2e-3c4884ec524e.png)](https://user-images.githubusercontent.com/3662109/208491336-54952b08-237e-4ef4-ad2e-3c4884ec524e.png)

- CPU time consumed by processes via Linux kernel interfaces.
- The majority of GitHub Enterprise service CPU time occurs here.
- On busy systems, `unicorn` consumes the most CPU time, followed by `babeld`, `git`, `git-daemon`, and `aqueduct`.

#### I/O operations (Read IOPS)

[![I/O operations (Read IOPS) graph](https://user-images.githubusercontent.com/3662109/46370731-3f2bb800-c63b-11e8-939e-4f0245e43209.png)](https://user-images.githubusercontent.com/3662109/46370731-3f2bb800-c63b-11e8-939e-4f0245e43209.png)

- `git-daemon`, and `babeld` read Input/Output Operations Per Second ( **IOPS** ) values are influenced by Git fetch / pull activity.
- `unicorn` read IOPS are influenced by web application or API `GET` requests.
- `aqueduct` read IOPS are from background jobs, such as regular repository maintenance and repacking.

#### I/O operations (Write IOPS)

[![I/O operations (Write IOPS) graph](https://user-images.githubusercontent.com/3662109/46370769-64b8c180-c63b-11e8-8d3c-2a281e5e1624.png)](https://user-images.githubusercontent.com/3662109/46370769-64b8c180-c63b-11e8-8d3c-2a281e5e1624.png)

- `git-daemon` write IOPS reflect Git push activity and is most often the largest consumer of write IOPS.
- `aqueduct` background jobs, such as search indexing, and repository repacking are also be a large consumer of write IOPS

#### Storage traffic (Read)

[![Storage traffic (Read) graph](https://user-images.githubusercontent.com/3662109/46370828-8ca82500-c63b-11e8-8c0c-fb1bb67fabd6.png)](https://user-images.githubusercontent.com/3662109/46370828-8ca82500-c63b-11e8-8c0c-fb1bb67fabd6.png)

- Read throughput trends are a counterpart to Read IOPS. These values used together can help determine if storage system read performance is as expected.
- User activity such as fetching, and retrieving API data will result in read activity.

#### Storage traffic (Written)

[![Storage traffic (Written)](https://user-images.githubusercontent.com/3662109/46370885-b1040180-c63b-11e8-9764-ca0bd6f824cb.png)](https://user-images.githubusercontent.com/3662109/46370885-b1040180-c63b-11e8-9764-ca0bd6f824cb.png)

- Write throughput trends are a counterpart to Write IOPS. These values used together can help determine if storage system read performance is as expected.
- Pushes, background repacks, and API `POST` operations are often the largest influencers of this graph.

### Authentication

The authentication graphs break down the rates at which users and applications are authenticating to the GitHub Enterprise appliance. We also track the protocol or service type such as Git or API for the authentications, which is useful in identifying broad user activity trends. The authentication graphs can help find interesting trends or timeframes to look at when diving deeper into authentication and API request logs.

#### Authentication Totals

[![Authentication Totals graph](https://user-images.githubusercontent.com/3662109/45724805-a8f09000-bb6c-11e8-8605-19495f0bb8d8.png)](https://user-images.githubusercontent.com/3662109/45724805-a8f09000-bb6c-11e8-8605-19495f0bb8d8.png)

- Displays which methods users are authenticating with, and if they are successful in those attempts.
- Large numbers of failures usually indicate misconfigured clients which are failing repeatedly.

#### Authentication Rate

[![Authentication Rate graph](https://user-images.githubusercontent.com/3662109/45724804-a8f09000-bb6c-11e8-9017-a700eed3bb14.png)](https://user-images.githubusercontent.com/3662109/45724804-a8f09000-bb6c-11e8-9017-a700eed3bb14.png)

- Large numbers of authentications per second can cause authentication worker saturation.
- Automated requests or "polling" can be identified by a flat baseline, or intervals of authentications which occur regularly, even during off-peak times such as weekends or holidays.
- Human user authentication trends typically follow a bell curve, more closely matching your organization's daily business hours.

### LDAP

LDAP graphs will only display data if [LDAP Authentication](https://docs.github.com/en/enterprise-server/admin/identity-and-access-management/using-ldap-for-enterprise-iam/using-ldap) is enabled on the GitHub Enterprise appliance. These graphs can help to identify slow responses from your LDAP server, as well as the overall volume of LDAP password based authentications.

#### LDAP authentications

[![LDAP authentications graph](https://user-images.githubusercontent.com/3662109/45724913-1e5c6080-bb6d-11e8-831d-74ebac6c9a79.png)](https://user-images.githubusercontent.com/3662109/45724913-1e5c6080-bb6d-11e8-831d-74ebac6c9a79.png)

- If any timeouts appear in the graph, GitHub Enterprise was unable to communicate with the LDAP server in time for an authentication request to take place.
- Failures indicate that users or clients are attempting to authenticate with an invalid LDAP username or password.
- Using a [Personal Access Token](https://docs.github.com/enterprise-server/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) authentication instead of username and password for users can help reduce the number and frequency of requests which rely on the LDAP server.

#### LDAP authentication response time

[![LDAP authentication response time graph](https://user-images.githubusercontent.com/3662109/45724912-1e5c6080-bb6d-11e8-833f-5de1c404dfbf.png)](https://user-images.githubusercontent.com/3662109/45724912-1e5c6080-bb6d-11e8-833f-5de1c404dfbf.png)

- Useful for tracking LDAP server performance trends, from the perspective of the GitHub Enterprise appliance.
- LDAP responses which take longer than 10 seconds will result in a timeout for the authentication request.

#### LDAP Sync Totals

[![LDAP Sync Totals graph](https://user-images.githubusercontent.com/3662109/45724911-1dc3ca00-bb6d-11e8-9a69-48bc2a562e0b.png)](https://user-images.githubusercontent.com/3662109/45724911-1dc3ca00-bb6d-11e8-9a69-48bc2a562e0b.png)

- Reflects the number of `user`, `team`, and net `new_members` records which were synchronized via the [LDAP Synchronization feature](https://docs.github.com/enterprise-server/admin/identity-and-access-management/using-ldap-for-enterprise-iam/using-ldap#enabling-ldap-sync), when the feature is enabled.

#### LDAP Sync Runtime

[![LDAP Sync Runtime graph](https://user-images.githubusercontent.com/3662109/45771533-38905000-bbfa-11e8-8fbc-0d4dcf8779bb.png)](https://user-images.githubusercontent.com/3662109/45771533-38905000-bbfa-11e8-8fbc-0d4dcf8779bb.png)

- If the runtime of team or user sync cycles exceeds the current LDAP Sync interval, the interval should be increased to allow completion before the next cycle.
- Long run times may indicate poor LDAP server performance, or suboptimal configuration of Domain Bases and restricted groups.

## Continue the conversation

There's more in ["Understanding your GHES graphs" part 2](https://github.com/orgs/community/discussions/43877). Please let us know if you have any questions in the comments!

**Comment by Madugu1990** [comment]
I need guidance and mentorship on how I will run this beautiful opportunity. Who can help please?

**Reply by tuves** [reply]
Welcome to the GitHub Community, @Madugu1990. Here are a few resources to get you started:

- [How I Got Started with GitHub 🚀](https://github.com/orgs/community/discussions/32792)
- [Getting Started with GitHub Part 1: Best practices for setting up your GitHub account #39082](https://github.com/community/community/discussions/39082)

**Reply by 0schweps0** [reply]
It's a good idea

**Reply by 6504425** [reply]
https://github.com/orgs/community/discussions/42146#discussioncomment-4645721

**Comment by tylerlawrence2023** [comment]
I can't find the "button"

**Reply by walkersevenonethreenineone** [reply]
Which?

**Reply by Raycroft1** [reply]
- _**
@**_[]()

**Comment by ghost** [comment]
Help

**Comment by ghost** [comment]
Ti rinngrazio tantissimo per la spiegazioje e per l'aiuto. Ti assicuro che fino a ieri notte il io compter andava alla grande fino a quando è accaduto quello che mi ha fatto perderesoldi, tempo dalla polizia postale e anche un device da ri settare, con tanti bug. ( ovviamwnt ho fatto tante foto fi molte persone che si sono acccanite con me nonostante io fossi impassibile, fino alle 20:00 ora italiana circa. avrei necessità di esporre l'accaduto e i reati a un amministratore in gamba e che possa cappirmi e aiutarmi. Consci qualcuno che possa dedicarmi qualche minuto nella privacy più assoluta?, Ancora grazie :)

**Comment by Junifa** [comment]
Nice

**Comment by MustafaAlsaeed00100** [comment]
I need Sam one do. Project for me
